// forge-model-derivative/src/model/FormatsFormats.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeModelDerivative) {
      root.ForgeModelDerivative = {};
    }
    root.ForgeModelDerivative.FormatsFormats = factory(root.ForgeModelDerivative.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';




  /**
   * The FormatsFormats model module.
   * @module model/FormatsFormats
   */

  /**
   * Constructs a new <code>FormatsFormats</code>.
   * @alias module:model/FormatsFormats
   * @class
   */
  var exports = function() {
    var _this = this;







  };

  /**
   * Constructs a <code>FormatsFormats</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/FormatsFormats} obj Optional instance to populate.
   * @return {module:model/FormatsFormats} The populated <code>FormatsFormats</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('svf')) {
        obj['svf'] = ApiClient.convertToType(data['svf'], ['String']);
      }
      if (data.hasOwnProperty('thumbnail')) {
        obj['thumbnail'] = ApiClient.convertToType(data['thumbnail'], ['String']);
      }
      if (data.hasOwnProperty('stl')) {
        obj['stl'] = ApiClient.convertToType(data['stl'], ['String']);
      }
      if (data.hasOwnProperty('step')) {
        obj['step'] = ApiClient.convertToType(data['step'], ['String']);
      }
      if (data.hasOwnProperty('iges')) {
        obj['iges'] = ApiClient.convertToType(data['iges'], ['String']);
      }
      if (data.hasOwnProperty('obj')) {
        obj['obj'] = ApiClient.convertToType(data['obj'], ['String']);
      }
    }
    return obj;
  }

  /**
   * @member {Array.<String>} svf
   */
  exports.prototype['svf'] = undefined;
  /**
   * @member {Array.<String>} thumbnail
   */
  exports.prototype['thumbnail'] = undefined;
  /**
   * @member {Array.<String>} stl
   */
  exports.prototype['stl'] = undefined;
  /**
   * @member {Array.<String>} step
   */
  exports.prototype['step'] = undefined;
  /**
   * @member {Array.<String>} iges
   */
  exports.prototype['iges'] = undefined;
  /**
   * @member {Array.<String>} obj
   */
  exports.prototype['obj'] = undefined;



  return exports;
}));
