// forge-model-derivative/src/model/JobIgesOutputPayloadAdvanced.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeModelDerivative) {
      root.ForgeModelDerivative = {};
    }
    root.ForgeModelDerivative.JobIgesOutputPayloadAdvanced = factory(root.ForgeModelDerivative.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';




  /**
   * The JobIgesOutputPayloadAdvanced model module.
   * @module model/JobIgesOutputPayloadAdvanced
   */

  /**
   * Constructs a new <code>JobIgesOutputPayloadAdvanced</code>.
   * Advanced options for &#x60;iges&#x60; type.
   * @alias module:model/JobIgesOutputPayloadAdvanced
   * @class
   */
  var exports = function() {
    var _this = this;





  };

  /**
   * Constructs a <code>JobIgesOutputPayloadAdvanced</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/JobIgesOutputPayloadAdvanced} obj Optional instance to populate.
   * @return {module:model/JobIgesOutputPayloadAdvanced} The populated <code>JobIgesOutputPayloadAdvanced</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('tolerance')) {
        obj['tolerance'] = ApiClient.convertToType(data['tolerance'], 'Number');
      }
      if (data.hasOwnProperty('surfaceType')) {
        obj['surfaceType'] = ApiClient.convertToType(data['surfaceType'], 'String');
      }
      if (data.hasOwnProperty('sheetType')) {
        obj['sheetType'] = ApiClient.convertToType(data['sheetType'], 'String');
      }
      if (data.hasOwnProperty('solidType')) {
        obj['solidType'] = ApiClient.convertToType(data['solidType'], 'String');
      }
    }
    return obj;
  }

  /**
   * Possible values are between `0` and `1`. By default it is set at 0.001.
   * @member {Number} tolerance
   * @default 0.001
   */
  exports.prototype['tolerance'] = 0.001;
  /**
   * Possible values are `bounded`, `trimmed` and `wireframe`. By default it is set to bounded surface.
   * @member {module:model/JobIgesOutputPayloadAdvanced.SurfaceTypeEnum} surfaceType
   * @default 'bounded'
   */
  exports.prototype['surfaceType'] = 'bounded';
  /**
   * Export the sheet body to IGES. `open`, `shell`, `surface` or `wireframe`. By default, it is set to `surface`.
   * @member {module:model/JobIgesOutputPayloadAdvanced.SheetTypeEnum} sheetType
   * @default 'surface'
   */
  exports.prototype['sheetType'] = 'surface';
  /**
   * Export the solid body to IGES `solid`, `surface` or `wireframe`. By default, it is set to `solid`.
   * @member {module:model/JobIgesOutputPayloadAdvanced.SolidTypeEnum} solidType
   * @default 'solid'
   */
  exports.prototype['solidType'] = 'solid';


  /**
   * Allowed values for the <code>surfaceType</code> property.
   * @enum {String}
   * @readonly
   */
  exports.SurfaceTypeEnum = {
    /**
     * value: "bounded"
     * @const
     */
    "bounded": "bounded",
    /**
     * value: "trimmed"
     * @const
     */
    "trimmed": "trimmed",
    /**
     * value: "wireframe"
     * @const
     */
    "wireframe": "wireframe"  };

  /**
   * Allowed values for the <code>sheetType</code> property.
   * @enum {String}
   * @readonly
   */
  exports.SheetTypeEnum = {
    /**
     * value: "open"
     * @const
     */
    "open": "open",
    /**
     * value: "surface"
     * @const
     */
    "surface": "surface",
    /**
     * value: "shell"
     * @const
     */
    "shell": "shell",
    /**
     * value: "wireframe"
     * @const
     */
    "wireframe": "wireframe"  };

  /**
   * Allowed values for the <code>solidType</code> property.
   * @enum {String}
   * @readonly
   */
  exports.SolidTypeEnum = {
    /**
     * value: "solid"
     * @const
     */
    "solid": "solid",
    /**
     * value: "surface"
     * @const
     */
    "surface": "surface",
    /**
     * value: "wireframe"
     * @const
     */
    "wireframe": "wireframe"  };


  return exports;
}));
