// forge-model-derivative/src/model/JobPayloadInput.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2016 
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 *
 */
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.ForgeModelDerivative) {
      root.ForgeModelDerivative = {};
    }
    root.ForgeModelDerivative.JobPayloadInput = factory(root.ForgeModelDerivative.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';




  /**
   * The JobPayloadInput model module.
   * @module model/JobPayloadInput
   */

  /**
   * Constructs a new <code>JobPayloadInput</code>.
   * Group of inputs
   * @alias module:model/JobPayloadInput
   * @class
   * @param urn {String} The design URN; returned when uploading the file to Forge The URN needs to be [Base64 (URL Safe) encoded](https://developer.autodesk.com/en/docs/model-derivative/v2/reference/http/job-POST/#id3). 
   */
  var exports = function(urn) {
    var _this = this;

    _this['urn'] = urn;


  };

  /**
   * Constructs a <code>JobPayloadInput</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/JobPayloadInput} obj Optional instance to populate.
   * @return {module:model/JobPayloadInput} The populated <code>JobPayloadInput</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('urn')) {
        obj['urn'] = ApiClient.convertToType(data['urn'], 'String');
      }
      if (data.hasOwnProperty('compressedUrn')) {
        obj['compressedUrn'] = ApiClient.convertToType(data['compressedUrn'], 'Boolean');
      }
      if (data.hasOwnProperty('rootFilename')) {
        obj['rootFilename'] = ApiClient.convertToType(data['rootFilename'], 'String');
      }
    }
    return obj;
  }

  /**
   * The design URN; returned when uploading the file to Forge The URN needs to be [Base64 (URL Safe) encoded](https://developer.autodesk.com/en/docs/model-derivative/v2/reference/http/job-POST/#id3). 
   * @member {String} urn
   */
  exports.prototype['urn'] = undefined;
  /**
   * Set this to `true` if the source file is compressed. If set to `true`, you need to define the `rootFilename`.
   * @member {Boolean} compressedUrn
   * @default false
   */
  exports.prototype['compressedUrn'] = false;
  /**
   * The root filename of the compressed file. Mandatory if the `compressedUrn` is set to `true`.
   * @member {String} rootFilename
   */
  exports.prototype['rootFilename'] = undefined;



  return exports;
}));
